[volume - 8] Decoupling with Kafka

rkdalsdn94 opened the pull request:
## 📌 Summary
<!--
    어떤 기능/이슈를 해결했는지 요약해주세요.
-->

### 요약
Kafka 기반 이벤트 파이프라인 구현

### 구현 사항
#### **1. Producer**
 - Transactional Outbox 패턴 구현
    - 비즈니스 로직과 동일한 트랜잭션 내에서 event_outbox 테이블 활용
    - 스케줄러(OutboxEventPublisher)가 주기적으로 PENDING 상태 이벤트를 Kafka로 전달
    - Kafka 장애 상황에도 이벤트 유실 방지
 - Partition Key 기반 순서 보장
    - aggregateId를 Partition Key로 사용하여 동일 엔티티의 이벤트 순서 보장
    - 예: 같은 상품의 좋아요 생성/취소 이벤트가 순서대로 처리됨
#### **2. Consumer**
 - Inbox 패턴으로 멱등성 보장
 - Batch Processing with Manual Commit
   - `max.poll.records=100`으로 배치 단위 처리
   - 배치 내 모든 메시지 처리 완료 후 수동 Offset Commit
   - 처리 중 실패한 메시지는 DLQ로 전송
 - Dead Letter Queue (DLQ)
   - 처리하지 못한 메시지를 dead_letter_queue 테이블에 저장
#### 3. 캐시 무효화 전략
 - 재고 소진 시 상품 캐시 무효화 (`OrderFacade`에서 `product.getStock() == 0`일 때 무효화)
#### 4. Embbed Kafka를 통한 카프카 학습 테스트
 - modules/kafka/src/test/java/com/loopers/kafka/learning/ 패키지에 학습 테스트 작성
   - 1. Auto Commit vs Manual Commit 비교
   - 2. earliest vs latest 동작 학습
   - 3. max.poll.interval.ms 타임아웃
   - 4. 배치 크기에 따른 성능 트레이드오프 테스트
   - 실제 카프카를 실행하는 방식이 아닌, **Spring 내부 Embbed Kafka로 테스트 진행**
   **간단한 설정 테스트라고 판단**했기 때문에 이 방식을 선택함

## 💬 Review Points
<!--
    리뷰어가 더 효과적인 리뷰를 할 수 있도록 도와주는 부분입니다.
    (1) 리뷰어가 중점적으로 봐줬으면 하는 부분
    (2) 고민했던 설계 포인트나 로직
    (3) 리뷰어가 확인해줬으면 하는 테스트 케이스나 예외 상황
    (4) 기타 리뷰어가 참고해야 할 사항
-->
### Q. `왜 이벤트 핸들링 테이블과 로그 테이블을 분리하는 걸까?` 에 대한 저의 생각입니다.
저는 `핸들링 테이블`과 `로그 테이블`을 각각 `inbox 테이블`과 `outbox 테이블`로 생각했습니다.

다음은 제가 생각하는 분리하는 이유 입니다.

#### 1. 두 테이블은 사용하는 목적이 다릅니다.

`inbox`는 이 이벤트를 이미 처리했는지에 대한 판단을 합니다.따라서 이벤트를 처리하기 직전에
중복된 이벤트인가를 살펴보고 중복 방지 목적이 끝나면 삭제하는 데이터 입니다. (ex. 10일이 지나면 삭제)

`outbox`는 이 이벤트를 MQ(여기선 Kafka)에 발행했는지에 대한 기록입니다.
따라서 이벤트 발행에 대해 발행했는지, 언제, 어떤 내용으로 발행했는지에 대한 목적을 가집니다.

#### 2. 데이터 보존 기간의 차이
`inbox`의 목적은 중복만 방지하면 되므로 **최근 이벤트만 필요**합니다.
하지만, `outbox`는 이벤트에 대해 감사 로그 역할도 겸하므로 **장기 보관이 필요**하다고 생각합니다.

---

### Q. 이번 구현에서 Exactly-Once를 보장하기 위해 적용한 방식입니다.

일단 **Exactly-Once**를 보장하려면 어떻게 할 수 있을지에 대한 내용입니다. (Kafka를 사용하는 상황)
####  1. Kafka Transactions (Kafka Native Exactly-Once)
  - Kafka 설정을 이용해 이벤트 중복 방지
  - enable.idempotence: true, transactional.id 등 설정

####  2. 비즈니스 레벨에서 Idempotency 구현
  - DB에서 비즈니스 고유키를 만들어 이벤트 중복 방지
  - 예: UNIQUE INDEX (user_id, product_id)

####  3. Upsert 방식
  - INSERT ... ON DUPLICATE KEY UPDATE 방식
  - 있으면 Update, 없으면 Insert

####  4. Inbox 패턴 사용
  - event_inbox 테이블을 만들어 event_id 기반 중복 체크
  - Inbox에 있는 이벤트는 무시, 없다면 처리 후 추가

#### 현재 선택한 방식
저는 `Inbox 패턴`은 선택했습니다. 이유로는 다음과 같습니다.

1. Kafka 방식의 문제점
외부 DB와 Kafka를 하나의 트랜잭션으로 어떻게 묶을 수 있을지에 대해 해결되지 않았습니다..
2. 비즈니스 레벨에서 Idempotencey의 문제점
비즈니스 키가 없는 경우 적용할 수 없다고 판단했습니다.
3. Upsert의 문제점
중복 이벤트 시 값이 2배로 증가(ex. 조회수)할 수 있다고 판단했습니다. (멱등성이 보장되지 않음)
 
4. 따라서 `Inbox` 테이블로 구현했습니다.
사실 처음에는 `Inbox`까지 필요할까 라는 마음으로 구현을 시작 했습니다.  
하지만, 루퍼스를 진행하는 과정 내에서는 이번 주에는 어떤 것을 배우고, 적용할 수 있을까에 대해 고민해보다가  
최대한 고민을 많이 담을 수 있는 방식으로 구현하는 방법을 선택했습니다.
Inbox 테이블은 주기적으로 삭제를 해야 되는데 **트래픽이 엄청 몰려 Inbox 테이블이 삭제될 때 downtime이 발생하면 어떡하지?**  등 추가적인 고민을 할 수 있는 방법을 선택했습니다.. 잘 한 건지는 모르겠습니다..!


## ✅ Checklist
<!--
    해당 작업이 완료되었는지 확인하기 위한 체크리스트입니다.
    리뷰어가 확인해야 할 사항을 포함합니다.
    계획중이나 아직 완료되지 않은 작업 또한 `TODO -` 로 작성해주세요.  

    ex.
    - [ ] 테스트 코드 포함
    - [ ] 불필요한 코드 제거
    - [ ] README or 주석 보강 (필요 시)
-->

<h3>🎾 Producer</h3>
<ul>
<li>[x]  도메인(애플리케이션) 이벤트 설계</li>
<li>[x]  Producer 앱에서 도메인 이벤트 발행 (catalog-events, order-events, 등)</li>
<li>[x]  <strong>PartitionKey</strong> 기반의 이벤트 순서 보장</li>
<li>[x]  메세지 발행이 실패했을 경우에 대해 고민해보기</li>
</ul>
<h3>⚾ Consumer</h3>
<ul>
<li>[x]  Consumer 가 Metrics 집계 처리</li>
<li>[x]  <code>event_handled</code> 테이블을 통한 멱등 처리 구현</li>
<li>[x]  재고 소진 시 상품 캐시 갱신</li>
<li>[x]  중복 메세지 재전송 테스트 → 최종 결과가 한 번만 반영되는지 확인</li>
</ul>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

# 릴리스 노트

**새 기능**
- 카프카 기반 이벤트 발행 시스템으로 신뢰성 있는 메시지 처리 지원
- 실패한 이벤트 관리를 위한 Dead Letter Queue 추가
- 중복 이벤트 자동 감지를 위한 Event Inbox 구현
- 상품별 메트릭(좋아요, 조회수, 판매량) 자동 추적

**리팩토링**
- 코드 구조 개선 및 캐시 무효화 최적화

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rkdalsdn94 committed:
feat: Kafka 이벤트 파이프라인을 위한 DB 스키마 추가

- event_outbox: Producer Outbox 패턴 (At Least Once 보장)
- event_inbox: Consumer Inbox 패턴 (멱등성 보장)
- dead_letter_queue: 실패 메시지 처리
- product_metrics: 실시간 집계 테이블 (좋아요/조회수/주문수/판매금액)

rkdalsdn94 committed:
feat: Kafka 설정 추가 및 KRaft 모드 적용

- KRaft 모드 활성화 (Zookeeper 제거)
- Producer: At Least Once 설정 (acks=all, idempotence=true)
- Consumer: 배치 처리 및 수동 오프셋 커밋
- 테스트 환경: auto.offset.reset=earliest 설정
- Outbox Publisher 스케줄러 활성화

rkdalsdn94 committed:
feat: Outbox 패턴 구현으로 At Least Once 발행 보장

- OutboxEventService: 이벤트를 Outbox 테이블에 저장
- OutboxEventPublisher: 스케줄러로 미발행 이벤트 조회 및 Kafka 발행
- At Least Once 보장: 트랜잭션 커밋 후 이벤트 발행
- 실패 시 재시도: 스케줄러가 주기적으로 미발행 이벤트 재발행

rkdalsdn94 committed:
feat: Kafka Consumer 구현 및 ProductMetrics 실시간 집계

Consumer 구현:
- CatalogEventConsumer: 좋아요/조회수 이벤트 처리
- OrderEventConsumer: 주문/결제 이벤트 처리
- Inbox 패턴으로 멱등성 보장 (중복 메시지 방지)
- 배치 처리로 성능 최적화
- DLQ(Dead Letter Queue)로 실패 메시지 처리

ProductMetrics 집계:
- 좋아요 수, 조회수, 주문 수, 판매 금액 실시간 집계
- 별도 테이블로 조회 성능 최적화

rkdalsdn94 committed:
feat: 재고 소진 시 상품 캐시 자동 갱신

- OrderFacade: 재고 차감 후 재고가 0이 되면 캐시 삭제
- ProductCacheService: Spring Cache 삭제 메서드 추가
- 다음 조회 시 DB에서 최신 데이터(재고 0)를 가져와 캐시 갱신

rkdalsdn94 committed:
test: Kafka Consumer 단위 테스트 추가

테스트 작성:
- CatalogEventConsumerTest: 좋아요/조회수 이벤트 처리 검증 (6개)
- OrderEventConsumerTest: 주문/결제 이벤트 처리 검증 (7개)
- EventInboxServiceTest: 멱등성 보장 검증
- ProductMetricsServiceTest: 집계 로직 검증
- ConsumerSmokeTest: Bean 생성 및 Context 로딩 검증

테스트 전략:
- Consumer 로직 단위 테스트 (Mock 사용)
- 실제 Kafka 없이 비즈니스 로직만 검증
- 문서로서의 테스트: 명확한 설명과 시나리오

rkdalsdn94 committed:
style: Import 최적화 및 코드 포맷팅

- IntelliJ Optimize Imports 적용
- 사용하지 않는 import 제거
- import 순서 정리 (static import 우선)

rkdalsdn94 committed:
test: Circuit Breaker 테스트 안정성 개선

- Retry와 Circuit Breaker 상호작용을 고려한 검증 수정
- times(6)에서 atLeast(6)으로 변경하여 재시도 로직 반영
- Mock 설정 누락으로 인한 NullPointerException 해결

rkdalsdn94 committed:
test: Kafka 설정 학습 테스트 추가

Embedded Kafka를 활용한 4가지 학습 테스트 구현:
- Experiment1: Auto Commit vs Manual Commit 동작 비교
- Experiment2: auto.offset.reset (earliest/latest) 학습
- Experiment3: max.poll.interval.ms 타임아웃 체험
- Experiment4: max.poll.records 배치 크기 트레이드오프

특징:
- @EmbeddedKafka 사용으로 Docker 불필요
- IntelliJ에서 바로 실행 가능
- 각 테스트는 독립적인 토픽 사용

looper88888 commented on apps/commerce-streamer/src/main/java/com/loopers/interfaces/consumer/CatalogEventConsumer.java:
```diff
@@ -0,0 +1,167 @@
+        int skipCount = 0;
+        int failCount = 0;
+
+        for (ConsumerRecord<Object, Object> record : records) {
+            try {
+                boolean processed = processEvent(record);
+                if (processed) {
+                    successCount++;
+                } else {
+                    skipCount++;  // 중복 이벤트
+                }
+            } catch (Exception e) {
+                failCount++;
+                log.error("❌ 이벤트 처리 실패 - partition: {}, offset: {}, key: {}, error: {}",
+                    record.partition(), record.offset(), record.key(), e.getMessage(), e);
+
+                // DLQ에 전송
+                sendToDLQ(record, e, 0);
+            }
+        }
+
+        // Offset 커밋 (배치 단위)
+        acknowledgment.acknowledge();
```
요거 실제로 문제입니다. 확인부탁드려요

rkdalsdn94 committed:
docs: 문서에서 내부 참조 경로 제거

README 파일에서 프로젝트 내부 참조를 일반적인 설명으로 수정
- performance-test/README.md
- modules/kafka/learning/README.md
